tests/functional/utils/convergence_analysis.py

```diff
@@ -0,0 +1,313 @@
+
+    Current support:
+        - Simplicial and Cartesian grids in 2d and 3d. Tensor grids are not supported.
+        - Static and time-dependent models.
+        - If the model is time-dependent, we assume that the ``TimeManager`` is
+          instantiated with a constant time step.
+
+    Parameters:
+        model_class: Model class for which the analysis will be performed.
+        model_params: Model parameters. We assume that it contains all parameters
+            necessary to set up a valid instance of :data:`model_class`.
+        levels: ``default=1``
+
+            Number of refinement levels associated to the convergence analysis.
+        in_space: ``default=True``
+
+            Whether convergence in space should be performed.
+        in_time:  ``default=False``
+
+            Whether convergence in time should be performed.
+        spatial_rate: ``default=1``
+
+            Rate at which the mesh size(s) should be refined. For example, use
+            ``spatial_rate=2`` for halving the mesh size(s) in-between levels.
+        temporal_rate: ``default=1``
+
+            Rate at which the time step size should be refined. For example,
+            use ``temporal_rate=2`` for halving the time step size in-between levels.
+
+    """
+
+    def __init__(
+        self,
+        model_class,
+        model_params: dict,
+        levels: int = 1,
+        in_space: bool = True,
+        in_time: bool = False,
+        spatial_rate: int = 1,
+        temporal_rate: int = 1,
+    ):
+        # Sanity checks
+        if not in_space and not in_time:
+            raise ValueError("At least one type of analysis should be performed.")
+
+        if not in_space and spatial_rate > 1:
+            warnings.warn("'spatial_rate' is not being used.")
+            spatial_rate = 1
+
+        if not in_time and temporal_rate > 1:
+            warnings.warn("'temporal_rate' is not being used.")
+            temporal_rate = 1
+
+        self.model_class = model_class
+        """Model class that should be used to run the simulations and perform the
+        convergence analysis.
+
+        """
+
+        self.levels: int = levels
+        """Number of levels of the convergence analysis."""
+
+        self.in_space: bool = in_space
+        """Whether a spatial analysis should be performed."""
+
+        self.in_time: bool = in_time
+        """Whether a temporal analysis should be performed."""
+
+        self.spatial_rate: int = spatial_rate
+        """Rate at which the mesh size should be refined. A value of ``2``
+        corresponds to halving the mesh size(s) between :attr:`levels`.
+
+        """
+
+        self.temporal_rate: int = temporal_rate
+        """Rate at which the time step size should be refined. A value of ``2``
+        corresponds to halving the time step size between :attr:`levels`.
+
+        """
+
+        # Initialize setup and retrieve spatial and temporal data
+        setup = model_class(model_params)
```

Comment:
I would copy model_params before passing them to the model class instantiation. You never know what prepare_simulation might do to them...

Reply:
Point taken.